File: quick-sort/python/quick_sort.py
import random


def sort(data):
    quick_sort(data, 0, len(data) - 1)


def quick_sort(data, start_idx, end_idx):
    if start_idx < end_idx:
        partition_idx = partition(data, start_idx, end_idx)

        quick_sort(data, start_idx, partition_idx - 1)
        quick_sort(data, partition_idx + 1, end_idx)


def partition(data, start_idx, end_idx):
    pivot = data[end_idx]

    i = start_idx
    for j in range(start_idx, end_idx):
        if data[j] < pivot:
            data[i], data[j] = data[j], data[i]

            i += 1

    data[i], data[end_idx] = data[end_idx], data[i]

    print(data)

    return i


if __name__ == "__main__":
    data = [random.randint(1, 20) for _ in range(20)]

    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)

File: quick-sort/golang/quick_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func sort(data []int) {
	quickSort(data, 0, len(data)-1)
}

func quickSort(data []int, startIdx, endIdx int) {
	if startIdx < endIdx {
		partitionIdx := partition(data, startIdx, endIdx)

		quickSort(data, startIdx, partitionIdx-1)
		quickSort(data, partitionIdx+1, endIdx)
	}
}

func partition(data []int, startIdx, endIdx int) int {
	pivot := data[endIdx]

	i := startIdx
	for j := startIdx; j < endIdx; j++ {
		if data[j] < pivot {
			data[j], data[i] = data[i], data[j]

			i++
		}
	}

	data[i], data[endIdx] = data[endIdx], data[i]

	fmt.Println(data)

	return i
}

func main() {
	data := make([]int, 20)

	for i := range 20 {
		data[i] = rand.Intn(20) + 1
	}

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}
